src/mpid/ch4/generic/am/mpidig_am_recv.h

```diff
@@ -140,6 +140,8 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_do_irecv(void *buf, MPI_Aint count, MPI_Data
         MPIDIG_rreq_dequeue(rank, tag, context_id, &MPIDI_global.unexp_list, MPIDIG_PT2PT_UNEXP);
 
     if (unexp_req) {
+        unexp_req->comm = comm;
+        MPIR_Comm_add_ref(comm);
```

Comment:
The concern is that adding reference count could be expensive. The main purpose of adding reference count is to prevent `comm` get released. Since we didn't have this reference count before and it works fine, it means either we don't really need comm to be alive or the comm is guaranteed to be alive by semantics. Either way, this particular reference counting probably is extra and can be avoided.

One way is to set `req->comm` to `NULL` in `MPIR_Request_complete`, thus preventing the extre release in `MPIR_Request_free`. But admittedly keeping track of which `comm` need be released at which point is tricky.

Another way is to keep the `comm` in a separate union filed. That way we can keep the usage/component isolated so it can be easier to track. The con is more code to manage.

Reply:
These references replace the references removed in this commit https://github.com/pmodels/mpich/pull/5304/commits/b746ad4a7375cd2e702e089b292eca074acc8547. They are still necessary to keep the communicator from being prematurely freed. It's just that before we refcounted based on enqueue/dequeue of the matching list, as opposed to request completion.

To be honest, we could probably come up with a counterexample that could show the previous refcounting could be broken. Something like dequeuing an RNDV or SSEND request after the user has freed the comm in the application could cause the communicator to go away before the CTS or ACK message is send back to the sender.

Reply:
I see. That makes sense

Reply:
This program fails for me with MPICH 3.4.1. I believe it is valid. Re-checking with my PR branch...
```
#include <mpi.h>
#include <stdlib.h>

int main(void) {
    MPI_Init(NULL, NULL);

    int rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    MPI_Comm comm;
    MPI_Comm_dup(MPI_COMM_WORLD, &comm);

    if (rank == 0) {
        MPI_Barrier(MPI_COMM_WORLD);
        MPI_Ssend(NULL, 0, MPI_INT, 1, 0, comm);
        MPI_Comm_free(&comm);
    } else {
        MPI_Request req;
        MPI_Irecv(NULL, 0, MPI_INT, 0, 0, comm, &req);
        MPI_Comm_free(&comm);
        MPI_Barrier(MPI_COMM_WORLD);
        MPI_Wait(&req, MPI_STATUS_IGNORE);
    }

    MPI_Finalize();

    return 0;
}
```

Reply:
Yeah, it works correctly with this branch. Now I have to think about backporting it 😅.